compiler/rustc_mir_dataflow/src/value_analysis.rs

```diff
@@ -0,0 +1,838 @@
+                values[vi] = value.clone();
+            }
+        });
+    }
+
+    pub fn flood_idx(&mut self, place: PlaceIndex, map: &Map) {
+        self.flood_idx_with(place, map, V::top())
+    }
+
+    /// This method assumes that the given places are not overlapping, and that we can therefore
+    /// copy all entries one after another.
+    pub fn assign_place_idx(&mut self, target: PlaceIndex, source: PlaceIndex, map: &Map) {
+        let StateData::Reachable(values) = &mut self.0 else { return };
+        // If both places are tracked, we copy the value to the target. If the target is tracked,
+        // but the source is not, we have to invalidate the value in target. If the target is not
+        // tracked, then we don't have to do anything.
+        if let Some(target_value) = map.places[target].value_index {
+            if let Some(source_value) = map.places[source].value_index {
+                values[target_value] = values[source_value].clone();
+            } else {
+                values[target_value] = V::top();
+            }
+        }
+        for target_child in map.children(target) {
+            // Try to find corresponding child and recurse. Reasoning is similar as above.
+            let projection = map.places[target_child].proj_elem.unwrap();
+            if let Some(source_child) = map.projections.get(&(source, projection)) {
+                self.assign_place_idx(target_child, *source_child, map);
+            } else {
+                self.flood_idx(target_child, map);
+            }
+        }
+    }
+
+    pub fn assign(&mut self, target: PlaceRef<'_>, result: ValueOrPlaceOrRef<V>, map: &Map) {
+        if let Some(target) = map.find(target) {
+            self.assign_idx(target, result, map);
+        } else {
+            // We don't track this place nor any projections, assignment can be ignored.
+        }
+    }
+
+    pub fn assign_idx(&mut self, target: PlaceIndex, result: ValueOrPlaceOrRef<V>, map: &Map) {
+        match result {
+            ValueOrPlaceOrRef::Value(value) => {
+                // First flood the target place in case we also track any projections (although
+                // this scenario is currently not well-supported by the API).
+                self.flood_idx(target, map);
+                let StateData::Reachable(values) = &mut self.0 else { return };
+                if let Some(value_index) = map.places[target].value_index {
+                    values[value_index] = value;
+                }
+            }
+            ValueOrPlaceOrRef::Place(source) => self.assign_place_idx(target, source, map),
+            ValueOrPlaceOrRef::Ref(source) => {
+                let StateData::Reachable(values) = &mut self.0 else { return };
+                if let Some(value_index) = map.places[target].value_index {
+                    values[value_index] = V::top();
+                }
+                if let Some(target_deref) = map.apply_elem(target, ProjElem::Deref) {
```

Comment:
Possibly add a comment like:

> If we know that the place we are referencing currently holds a particular constant, then mark this as being the value at the place created by dereferencing the target of the assignment. This is correct because we checked above that the type of the referenced place is `Freeze` and stacked borrows provides that the value behind a `&Freeze` cannot be changed between two uses of the reference.

Reply:
I realized this slightly differently, by only tracking places that are `Freeze`. I know that this may prevent optimizations in rare cases, but this was easier to realize for now. Let me know if you think that this is unsound.